File: src/utils/SignValidator.java
package utils;

import java.util.Stack;

public class SignValidator {
    
    public boolean isValid(String signos){
        Stack<Character> stack = new Stack<>();
        for (Character character : signos.toCharArray()) {
            // character -> {
            // character -> }
            // 1 --> Comprobar si es de apertura o cierre
            if(getIsApertura(character)){
                stack.push(character);
            }else{
                if(stack.isEmpty()){
                    return false;
                }
                char ultimo = stack.pop();
                //ultimo es del mismo tipo que character?
                boolean mismoTipo = isTipoIgual(character, ultimo);
                if (mismoTipo == false){
                    return false;
                }
            }
        }
        return stack.isEmpty();
    }

    private boolean getIsApertura(char c){
        return c=='{' || c=='(' || c=='[';
    }

    private boolean isTipoIgual(char c, char ultimo){
        return (c==')' && ultimo=='(') || (c=='}' && ultimo=='{')  || (c==']' && ultimo=='[');
    }
}

File: bin/python/SignValidator.py
class SignValidator:

    def is_valid(self, signos: str) -> bool:
        stack = []

        for character in signos:
            # character -> {
            # character -> }
            # 1 --> Comprobar si es de apertura o cierre
            if self.is_apertura(character):
                stack.append(character)
            else:
                if len(stack) == 0:
                    return False

                ultimo = stack.pop()
                # ultimo es del mismo tipo que character?
                mismo_tipo = self.is_tipo_igual(character, ultimo)

                if mismo_tipo is False:
                    return False

        return len(stack) == 0

    def is_apertura(self, c: str) -> bool:
        return c == '{' or c == '(' or c == '['

    def is_tipo_igual(self, c: str, ultimo: str) -> bool:
        return (c == ')' and ultimo == '(') or \
               (c == '}' and ultimo == '{') or \
               (c == ']' and ultimo == '[')
